clang/lib/AST/Interp/ByteCodeExprGen.h

```diff
@@ -478,6 +480,43 @@ template <class Emitter> class SourceLocScope final {
   bool Enabled = false;
 };
 
+template <class Emitter> class StoredOpaqueValueScope final {
+public:
+  StoredOpaqueValueScope(ByteCodeExprGen<Emitter> *Ctx) : Ctx(Ctx) {}
+
+  bool VisitAndStoreOpaqueValue(const OpaqueValueExpr *Ove) {
+    assert(Ove && "OpaqueValueExpr is a nullptr!");
+    assert(!Ctx->OpaqueExprs.contains(Ove) &&
+           "OpaqueValueExpr already stored!");
+
+    std::optional<PrimType> CommonTy = Ctx->classify(Ove);
+    std::optional<unsigned> LocalIndex = Ctx->allocateLocalPrimitive(
+        Ove, *CommonTy, Ove->getType().isConstQualified());
+    if (!LocalIndex)
+      return false;
+
+    if (!Ctx->visit(Ove))
+      return false;
+
+    if (!Ctx->emitSetLocal(*CommonTy, *LocalIndex, Ove))
+      return false;
+
+    Ctx->OpaqueExprs.insert({Ove, *LocalIndex});
+    StoredValues.emplace_back(Ove);
+
+    return true;
+  }
+
+  ~StoredOpaqueValueScope() {
+    for (const auto *SV : StoredValues)
+      Ctx->OpaqueExprs.erase(SV);
+  }
+
+private:
+  ByteCodeExprGen<Emitter> *Ctx;
+  std::vector<const OpaqueValueExpr *> StoredValues;
```

Comment:
Why is this vector necessary?

Reply:
I kept in mind the scenario, when we might have nested `StoredOpaqueValueScope`s. Somehow we need to keep track of which `OpaqueValueExpr` should be removed from `Ctx->OpaqueExprs`, when we exit the scope.

Do you suggest keeping everything in `Ctx->OpaqueExprs` during the whole lifetime of the bytecode generator?

Reply:
Well you don't ever reuse the same `StoredOpaqueValueScope`, do you? The scope should add the value in its constructor and remove it in its destructor. So as long as we don't evaluate the same `OpaqueExpr` twice, we should be fine and no vector is needed.

Reply:
My idea was that multiple `OpaqueValueExpr` can be in the same scope, but now I removed the scope completely.